### PR TITLE
ISBN Simulation (Jira #6)

### PR DESCRIPTION
**Pull request by JenOmega**
A dirty way to simulate ISBN scanning with a little loading/error message (https://candid-eng.atlassian.net/browse/BOOK-6)

### CODE REVIEW

File: src-ui/lib/components/BooksTable.svelte [hunk 1]
```diff
@@ -15,12 +15,13 @@
     hasRead: true,
   }
 
+
```
Comment (SilverMaiden) [inline src-ui/lib/components/BooksTable.svelte]: Did prettier do this?

File: src-ui/lib/components/BooksTable.svelte [hunk 2]
```diff
@@ -29,8 +30,9 @@
       qty: number
     }
   }
+  let promise: Promise<void> | undefined;
```
Comment (SilverMaiden) [inline src-ui/lib/components/BooksTable.svelte:33]: Let's rename the variable to simulateScanPromise or something similar?

File: src-ui/lib/components/BooksTable.svelte [hunk 3]
```diff
@@ -56,7 +58,18 @@
   }
 </script>
 
-<svelte:document on:scan={handleScan} use:listenForBarcodes />
+
+<button on:click={() => onScan.simulate(document, '1234567890123')}>
```
Comment (SilverMaiden) [inline src-ui/lib/components/BooksTable.svelte]: Should we disable the button if there's a scan simulation in progress? Also, I'd move the simulate functionality out into a separate function, and then pass that directly to the onClick. That way, we can also write a unit test for it.

File: src-ui/lib/components/BooksTable.svelte [hunk 4]
```diff
@@ -56,7 +58,18 @@
...
+	Simulate ISBN
+</button>
+
+{#await promise}
```
Comment (SilverMaiden) [inline src-ui/lib/components/BooksTable.svelte]: I might move the loading state out into its' own variable (isLoading) with a default value of `false`, and then update that value based on where in the process the addBook promise is. Having an explicit `isLoading` variable is easier to manage/reason about than relying on the promise here, imo.

File: src-ui/lib/components/BooksTable.svelte [hunk 5]
```diff
@@ -15,12 +15,13 @@
...
   }
 
+
   if (books.value.length == 0) {
     books.add(initialBook)
   }
 
   const addBook = async (isbn: string): Promise<void> => {
-    books.add(await getByISBN(isbn))
+      books.add(await getByISBN(isbn))
```
Comment (SilverMaiden) [inline src-ui/lib/components/BooksTable.svelte]: Same here - do we need the whitespace? If prettier added it, then ignore this.

**Review by SilverMaiden — COMMENTED** [review]
Ok taking a first pass! I'm pretty new to PR reviews, so take every comment with a grain of salt 🙏 Left some comments, most important part would be to write a unit test for the simulate functionality. @rkofman, feel free to veto any of these comments.

File: src-ui/lib/components/BooksTable.svelte (hunk 4 above)
Comment (rkofman) [reply src-ui/lib/components/BooksTable.svelte]: The `#await` block is a built-in Svelte pattern. It's not worse than using `isLoading`. The only question is whether there are other things that need to rely on the state; in which case we can expose that state. I would explore existing open source svelte code on Github to see if there are patterns to learn from.

File: src-ui/lib/components/BooksTable.svelte (hunk 3 above)
Comment (JenOmega) [reply src-ui/lib/components/BooksTable.svelte]: You're probably right about disabling the button - I'll fix that. 
the function (simulate) is already included in the scan library itself.

**Review by rkofman — APPROVED** [review]
I'm not sure why Github is showing so many changes. A diff from `main` is a single file; which I reviewed and looks good. Going to merge.

**Pull request merged**